Win32-network/src/System/Win32/Async/IOManager.hsc

```diff
@@ -0,0 +1,284 @@
+--
+newtype IOCompletionPort = IOCompletionPort HANDLE
+  deriving Show
+
+closeIOCompletionPort :: IOCompletionPort -> IO ()
+closeIOCompletionPort (IOCompletionPort iocp) = Win32.closeHandle iocp
+
+-- | Windows documentation:
+--
+createIOCompletionPort
+    :: DWORD  -- ^ number of concurrent threads
+    -> IO IOCompletionPort
+createIOCompletionPort concurrentThreads
+    = IOCompletionPort <$>
+        Win32.failIfNull "createIoCompletionPort"
+          (c_CreateIoCompletionPort
+            Win32.iNVALID_HANDLE_VALUE
+            nullPtr
+            nullPtr
+            concurrentThreads)
+
+foreign import ccall unsafe "windows.h CreateIoCompletionPort"
+    c_CreateIoCompletionPort :: HANDLE -> HANDLE -> Ptr () -> DWORD -> IO HANDLE
+
+-- | Associate with I/O completion port.  This can be used multiple times on
+-- a file descriptor.
+--
+associateWithIOCompletionPort :: Either HANDLE Socket
+                              -> IOCompletionPort
+                              -> IO ()
+associateWithIOCompletionPort (Left h) iocp = do
+    Win32.failIfFalse_ "associateWithIOCompletionPort" (c_AssociateHandle h iocp)
+associateWithIOCompletionPort (Right sock) iocp =
+    Socket.withFdSocket sock $ \fd -> do
+      Win32.failIfFalse_ "associateWithIOCompletionPort" (c_AssociateSocket fd iocp)
+
+foreign import ccall safe "HsAssociateHandle"
+    c_AssociateHandle :: HANDLE           -- ^ handle
+                      -> IOCompletionPort -- ^ I/O completion port
+                      -> IO BOOL
+
+foreign import ccall safe "HsAssociateSocket"
+    c_AssociateSocket :: CInt             -- ^ Socket descriptor
+                      -> IOCompletionPort -- ^ I/O completion port
+                      -> IO BOOL
+
+-- | Create an 'IOCompletionPort' and start a I/O manager thread (bound thread).
+-- The IOManager thread will only run for the duration of the callback.
+--
+-- TODO: add a tracer which logs when `dequeueCompletionPackets' errors
+--
+withIOManager :: (IOCompletionPort -> IO r) -> IO r
+withIOManager k =
+    bracket
+        (createIOCompletionPort maxBound)
+        closeIOCompletionPort
+        $ \iocp -> do
+          _ <- forkOS (void $ dequeueCompletionPackets iocp)
+          k iocp
+
+
+data IOCompletionException
+    = NullPointerException
+    | NullOverlappedPointer
+    | IOCompoletionError !Win32.ErrCode
+  deriving Show
+
+instance Exception IOCompletionException
+
+-- | I/O manager which handles completions of I/O operations.  It should run on
+-- a bound thread.  It dereferences the stable pointed which was allocated by
+-- `readAsync` or `writeAsync`, and puts the results of the operation in the
+-- 'MVar' which it points to.  This wakes up the 'waitForCompletion' function.
+--
+-- To safely shutdown the I/O manager, it's enough to close the port's 'HANDLE'.
+-- This should be done after all the associated handles are closed.
+--
+-- The underlaying 'c' code is using a fixed completion key which is not exposed
```

Comment:
underlaying/underlying

If we eliminate the C code entirely, the magic completion key will be used here instead of course.